x/gov/abci.go

```diff
@@ -0,0 +1,117 @@
+package gov
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+
+	"github.com/neutron-org/neutron/x/gov/keeper"
+)
+
+// EndBlocker called at the end of every block, processing proposals
+// same as the vanilla gov EndBlocker, except for we replace the `Tally`
+// function with our own implementation
+func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
+	defer telemetry.ModuleMeasureSince(govtypes.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
+	logger := keeper.Logger(ctx)
+
+	// delete inactive proposal from store and its deposits
+	keeper.IterateInactiveProposalsQueue(ctx, ctx.BlockHeader().Time, func(proposal govtypes.Proposal) bool {
+		keeper.DeleteProposal(ctx, proposal.ProposalId)
+		keeper.DeleteDeposits(ctx, proposal.ProposalId)
+
+		// called when proposal become inactive
+		keeper.AfterProposalFailedMinDeposit(ctx, proposal.ProposalId)
+
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				govtypes.EventTypeInactiveProposal,
+				sdk.NewAttribute(govtypes.AttributeKeyProposalID, fmt.Sprintf("%d", proposal.ProposalId)),
+				sdk.NewAttribute(govtypes.AttributeKeyProposalResult, govtypes.AttributeValueProposalDropped),
+			),
+		)
+
+		logger.Info(
+			"proposal did not meet minimum deposit; deleted",
+			"proposal", proposal.ProposalId,
+			"title", proposal.GetTitle(),
+			"min_deposit", keeper.GetDepositParams(ctx).MinDeposit.String(),
+			"total_deposit", proposal.TotalDeposit.String(),
+		)
+
+		return false
+	})
+
+	// fetch active proposals whose voting periods have ended (are passed the block time)
+	keeper.IterateActiveProposalsQueue(ctx, ctx.BlockHeader().Time, func(proposal govtypes.Proposal) bool {
+		var tagValue, logMsg string
+
+		passes, burnDeposits, tallyResults := keeper.Tally(ctx, proposal) // our custom implementation of tally logics
+
+		if burnDeposits {
+			keeper.DeleteDeposits(ctx, proposal.ProposalId)
+		} else {
+			keeper.RefundDeposits(ctx, proposal.ProposalId)
+		}
+
+		if passes {
```

Comment:
(note to self) this is a 0.45 version of governance, we'll need to patch this module to 0.46 if we choose to switch sdks